### PR TITLE
Windows2merge

### PR DESCRIPTION
**Pull request by philipatl**
#### Description
*Changes proposed in this pull request.*

#### Test plan
*Outline the test plan used to test this change before merging it.*

> Fixes #.

### CODE REVIEW

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "f280a9af27c401216150f26a2d9332bc1a0027af"
  }
]
```

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "db770341b789f8ec19b0fc886fe149ec8d3e9039"
  }
]
```

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "828b6f6310b55622dfc9631e61a47c1645c93436"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "9ff3d5afe69018730a947b81689ecea16b5790e0"
  }
]
```

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "828b6f6310b55622dfc9631e61a47c1645c93436"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "be70776145fb553ac182b30e71e094c09ca53896"
  }
]
```

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "ab450bb28b370ff267fba8f1c445f24ebf081665"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "be70776145fb553ac182b30e71e094c09ca53896"
  }
]
```

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "bc9e3d94bc5f8037b169beaa23738587437d7ebd"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "e3de664a7dacbdbfd58a483714a8decfad8bd5e2"
  }
]
```

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "8af40f9cb1c340d9b7e2ac3a581e89f6db6c7e23"
  },
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "ad99ff5162929b206d28213a4feced765c161c88"
  }
]
```

File: controller/internal/enforcer/applicationproxy/protomux/protomux.go [hunk 1]
```diff
@@ -175,9 +175,11 @@ func (m *MultiplexedListener) Serve(ctx context.Context) error {
 		case <-m.shutdown:
 			return nil
 		default:
+
 			c, err := m.root.Accept()
 			if err != nil {
-				return err
+				zap.L().Error("error from Accept", zap.Error(err))
```
Comment (mheese) [inline controller/internal/enforcer/applicationproxy/protomux/protomux.go:189]: please add a code comment here why we are not failing anymore (break instead of return) (Amit told to me that this is an intermittent error on Windows and needs retry, so that explanation would be nice to have)

File: controller/internal/enforcer/dnsproxy/dns_windows.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
+// +build windows
```
Comment (mheese) [inline controller/internal/enforcer/dnsproxy/dns_windows.go]: not necessary, the file ends in `_windows.go`

File: controller/internal/enforcer/dnsproxy/dns_osx.go [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// +build !linux darwin
+// +build darwin
```
Comment (mheese) [inline controller/internal/enforcer/dnsproxy/dns_osx.go]: file should end in `_darwin.go`, then you can remove the build line completely

File: controller/internal/enforcer/nfqdatapath/autoport_linux.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
+// +build linux
```
Comment (mheese) [inline controller/internal/enforcer/nfqdatapath/autoport_linux.go]: again, file ends in `_linux.go`, doesn't need this line

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "8af40f9cb1c340d9b7e2ac3a581e89f6db6c7e23"
  },
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "ad99ff5162929b206d28213a4feced765c161c88"
  }
]
```

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "20ee62bddb0d966552c9166b468af245a59cfac1"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "db87f76bba3c8a853e12ec84d70f96f7870f5f8f"
  }
]
```

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "e1df1595776aea7a3f9d4dc05d926dfc94169042"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "eafec8cfd3c89989c6628fb71913df4a152d5186"
  }
]
```

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "e5508908e26db2d811d5e81b038aa7a33346e4b5"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "fe82b52f91f3d7f558610b96f4eabbad8421f7c8"
  }
]
```

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "e5508908e26db2d811d5e81b038aa7a33346e4b5"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "7d9033460e1abb5367a020aa5d58d677cc13b8cf"
  }
]
```

File: controller/internal/enforcer/dnsproxy/dns_windows.go (hunk 2 above)
Comment (philipatl) [reply controller/internal/enforcer/dnsproxy/dns_windows.go]: there is already a dns.go which builds for linux

File: controller/internal/enforcer/nfqdatapath/autoport_linux.go (hunk 4 above)
Comment (philipatl) [reply controller/internal/enforcer/nfqdatapath/autoport_linux.go]: this reminds me that there is no osx/darwin build for these functions, so let me add that.

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "e5508908e26db2d811d5e81b038aa7a33346e4b5"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "ca7075d88f37ce395ba92405f4799f333b20614a"
  }
]
```

File: controller/internal/enforcer/applicationproxy/markedconn/markedconn.go [hunk 5]
```diff
@@ -178,102 +190,6 @@ func (l ProxiedListener) Close() error {
...
-
-	localIPString, _, err := net.SplitHostPort(conn.LocalAddr().String())
-	if err != nil {
-		return nil, 0, err
-	}
-
-	localIP := net.ParseIP(localIPString)
-	return getOriginalDestInternal(rawconn, localIP.To4() != nil, syscall.Syscall6)
-}
-
```
Comment (dstiliadis) [inline controller/internal/enforcer/applicationproxy/markedconn/markedconn.go:276]: Why is this gone? What's the replacement?

File: controller/internal/enforcer/applicationproxy/markedconn/nativedata.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+// NativeDataControl dummy impl
+type NativeDataControl struct {
+}
+
+// NewNativeDataControl returns initialized NativeDataControl
+func NewNativeDataControl() *NativeDataControl {
+	return &NativeDataControl{}
+}
+
```
Comment (dstiliadis) [inline controller/internal/enforcer/applicationproxy/markedconn/nativedata.go]: Not clear to me what "NativeData" means. Can we either document it or make the name more clear?

File: controller/internal/enforcer/nfqdatapath/afinetrawsocket/afinetrawsocket.go [hunk 7]
```diff
@@ -41,7 +41,7 @@ const (
 
 // SocketWriter interface exposes an interface to write and close sockets
 type SocketWriter interface {
-	WriteSocket(buf []byte, version packet.IPver) error
+	WriteSocket(buf []byte, version packet.IPver, data interface{}) error
```
Comment (dstiliadis) [inline controller/internal/enforcer/nfqdatapath/afinetrawsocket/afinetrawsocket.go]: I would prefer no generic interface definitions in the methods. Any reason we actually need an interface and we cannot define a datatype.

File: controller/internal/enforcer/nfqdatapath/afinetrawsocket/afinetrawsocket_windows.go [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+// CreateSocket returns a handle to SocketWriter interface
+func CreateSocket(mark int, deviceName string) (SocketWriter, error) {
+	return &rawsocket{}, nil
+}
+
+func (sock *rawsocket) WriteSocket(buf []byte, version packet.IPver, data interface{}) error {
+	w, _ := data.(*WindowsPacketMetadata)
+	if w == nil {
```
Comment (dstiliadis) [inline controller/internal/enforcer/nfqdatapath/afinetrawsocket/afinetrawsocket_windows.go]: This needs a type check (and thus my problem with interfaces) 

w, ok := data.(*WindowsPacketMetadata)
if !ok {
...
}

File: controller/internal/enforcer/nfqdatapath/datapath_tcp.go [hunk 9]
```diff
@@ -384,6 +384,9 @@ func (d *Datapath) processApplicationSynAckPacket(tcpPacket *packet.Packet, cont
 			zap.L().Debug("Failed to remove cache entries")
 		}
 
+		if err := d.ignoreFlow(tcpPacket, tcpPacket.WindowsMetadata); err != nil {
+			zap.L().Error("Failed to ignore flow", zap.Error(err))
+		}
 		if err := d.conntrack.UpdateApplicationFlowMark(
```
Comment (dstiliadis) [inline controller/internal/enforcer/nfqdatapath/datapath_tcp.go:390]: Seems to me that we don't need to pass both tcpPacket and tcpPacket.WindowsMedata , since tcpPacket has that already.

File: controller/internal/enforcer/nfqdatapath/nfq_windows.go [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+			// process udp packet
+			if packetType == packet.PacketTypeNetwork {
+				udpConn, processError = d.ProcessNetworkUDPPacket(parsedPacket)
+			} else {
+				udpConn, processError = d.ProcessApplicationUDPPacket(parsedPacket)
+			}
+		} else {
+			processError = fmt.Errorf("invalid ip protocol: %d", parsedPacket.IPProto())
+		}
+
```
Comment (dstiliadis) [inline controller/internal/enforcer/nfqdatapath/nfq_windows.go:87]: This logic seems a little convoluted, and I am not sure how it will work. 

If err!=nil, we print it, but not drop it or return. processError remains nil, and thus the subsequent if statement is not met. 

I would suggest we explicitly define on error and then return or drop (or whatever we need to do ), and then simplify the nested IFs.

File: controller/internal/enforcer/applicationproxy/markedconn/markedconn.go (hunk 5 above)
Comment (philipatl) [reply controller/internal/enforcer/applicationproxy/markedconn/markedconn.go:276]: getOriginalDestPlatform

File: controller/internal/enforcer/nfqdatapath/nfq_windows.go [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+					tcpConn: nil,
+					udpConn: udpConn,
+					err:     processError,
+					network: packetType == packet.PacketTypeNetwork,
+				})
+			}
+			// drop packet by not forwarding it
+			return 0
+		}
+
```
Comment (dstiliadis) [inline controller/internal/enforcer/nfqdatapath/nfq_windows.go:111]: Please replace with a switch statement for better readability.

File: controller/internal/enforcer/nfqdatapath/nfq_windows.go [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+		} else if parsedPacket.IPProto() == packet.IPProtocolUDP {
+			d.collectUDPPacket(&debugpacketmessage{
+				Mark:    mark,
+				p:       parsedPacket,
+				tcpConn: nil,
+				udpConn: udpConn,
+				err:     nil,
+				network: packetType == packet.PacketTypeNetwork,
+			})
+		}
```
Comment (dstiliadis) [inline controller/internal/enforcer/nfqdatapath/nfq_windows.go:151]: This if-then-else seems to be repeated multiple times and it makes it hard to follow. Any chance we can consolidate the reporting steps in a function and avoid this.

File: controller/internal/supervisor/iptablesctrl/acls_windows.go [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+	winRuleSpec2.GroupId = 0
+	winRuleSpec2.Action = 0
+	if reflect.DeepEqual(winRuleSpec1, winRuleSpec2) {
+		winRuleSpec1.Log = true
+		winRuleSpec1.LogPrefix = logPrefix
+		winRuleSpec1.GroupId = groupId
+		winRuleSpec1.Action = action
+		return winRuleSpec1
+	}
+	return nil
```
Comment (dstiliadis) [inline controller/internal/supervisor/iptablesctrl/acls_windows.go:203]: That gives me a little pain, because I am not even sure it always works and it is actually very slow. If we can find a way to avoid using reflect.DeepEqual I would prefer it.

File: controller/internal/supervisor/iptablesctrl/acls_windows.go [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+		}
+		logPrefix = winRuleSpec1.LogPrefix
+		groupId = winRuleSpec1.GroupId
+		if !winRuleSpec1.Log {
+			return nil
+		}
+	} else if !winRuleSpec2.Log {
+		return nil
+	}
+
```
Comment (dstiliadis) [inline controller/internal/supervisor/iptablesctrl/acls_windows.go:186]: I would avoid the if-then-else 

if !winRuleSpec2.Log {
  return nil 
}

Any place we have if then , else if .. should be optimized to a simpler structure or its impossible to follow.

File: controller/internal/supervisor/iptablesctrl/acls_windows.go [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+			// to that of the PU to discriminate
+			if winRuleSpec.Protocol == packet.IPProtocolTCP {
+				winRuleSpec.MatchDstPort, _ = winipt.ParsePortString(cfg.TCPPorts)
+			} else if winRuleSpec.Protocol == packet.IPProtocolUDP {
+				winRuleSpec.MatchDstPort, _ = winipt.ParsePortString(cfg.UDPPorts)
+			}
+		}
+	} else {
+		return nil, fmt.Errorf("unexpected Windows PU: %v", cfg.PUType)
+	}
```
Comment (dstiliadis) [inline controller/internal/supervisor/iptablesctrl/acls_windows.go:232]: Please replace with a switch statement.

File: controller/internal/windows/iptables_windows.go [hunk 16]
```diff
@@ -0,0 +1,337 @@
...
+				if ms.MatchSetSrcPort || ms.MatchSetDstPort {
+					rulespec += ","
+				}
+			} else if ms.MatchSetDstIp {
+				rulespec += "dstIP"
+				if ms.MatchSetSrcPort || ms.MatchSetDstPort {
+					rulespec += ","
+				}
+			}
+			if ms.MatchSetSrcPort {
```
Comment (dstiliadis) [inline controller/internal/windows/iptables_windows.go:103]: What happens if none matches?

File: controller/internal/windows/iptables_windows.go [hunk 17]
```diff
@@ -0,0 +1,337 @@
...
+	case frontman.FilterActionAllow:
+		rulespec += "-j ACCEPT "
+	case frontman.FilterActionBlock:
+		rulespec += "-j DROP "
+	case frontman.FilterActionProxy:
+		rulespec += fmt.Sprintf("-j REDIRECT --to-ports %d ", winRuleSpec.ProxyPort)
+	case frontman.FilterActionNfq:
+		rulespec += fmt.Sprintf("-j NFQUEUE -j MARK %d ", winRuleSpec.Mark)
+	}
+	if winRuleSpec.Log {
```
Comment (dstiliadis) [inline controller/internal/windows/iptables_windows.go:123]: Default action?

File: monitor/extractors/windows.go [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+	options := policy.OptionsType{}
+	for index, s := range event.Services {
+		if s.Port != 0 && s.Ports == nil {
+			if pspec, err := portspec.NewPortSpec(s.Port, s.Port, nil); err == nil {
+				event.Services[index].Ports = pspec
+				event.Services[index].Port = 0
+			} else {
+				return nil, fmt.Errorf("Invalid Port Spec %s", err)
+			}
+		}
```
Comment (dstiliadis) [inline monitor/extractors/windows.go:67]: we prefer the style 

if pspec, err := ... ; err != nil {
return 
}

continue with the non-error case.

**Review by dstiliadis — CHANGES_REQUESTED** [review]
see inline

File: controller/internal/enforcer/applicationproxy/markedconn/nativedata.go (hunk 6 above)
Comment (philipatl) [reply controller/internal/enforcer/applicationproxy/markedconn/nativedata.go]: Coming from the Windows world, I was using 'native' to mean "unmanaged" or outside the Go runtime/type system domain. I'll change it.

Comment (philipatl) [reply controller/internal/enforcer/applicationproxy/markedconn/nativedata.go]: Could be "UntypedData", "UnsafeData", "PlatformData". I'll try PlatformData unless there's another suggestion.

File: controller/internal/enforcer/nfqdatapath/nfq_windows.go (hunk 10 above)
Comment (philipatl) [reply controller/internal/enforcer/nfqdatapath/nfq_windows.go:87]: I believe this logic is similar to that of linux/master. Can you verify?

Comment (philipatl) [reply controller/internal/enforcer/nfqdatapath/nfq_windows.go:87]: https://github.com/aporeto-inc/trireme-lib/blob/master/controller/internal/enforcer/nfqdatapath/nfq_linux.go#L77

File: controller/internal/enforcer/nfqdatapath/afinetrawsocket/afinetrawsocket.go (hunk 7 above)
Comment (philipatl) [reply controller/internal/enforcer/nfqdatapath/afinetrawsocket/afinetrawsocket.go]: The metadata is stored on the packet.Packet structure, and it's an interface{} there following similar conventions for other members on that struct (due to circular reference?). But I can clean up this interface and make it take a concrete type.

File: monitor/extractors/windows.go (hunk 18 above)
Comment (philipatl) [reply monitor/extractors/windows.go:67]: pspec needs to be used in the non-error case and so can't be defined local to the if block. but I can restructure to check err and return early.

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "e5508908e26db2d811d5e81b038aa7a33346e4b5"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "8d9b47629b4427e1434daa5fa41c9fb8ef92c114"
  }
]
```

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "e5508908e26db2d811d5e81b038aa7a33346e4b5"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "8d9b47629b4427e1434daa5fa41c9fb8ef92c114"
  }
]
```

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "e5508908e26db2d811d5e81b038aa7a33346e4b5"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "8d9b47629b4427e1434daa5fa41c9fb8ef92c114"
  }
]
```

File: controller/internal/enforcer/nfqdatapath/afinetrawsocket/afinetrawsocket_windows.go (hunk 8 above)
Comment (philipatl) [reply controller/internal/enforcer/nfqdatapath/afinetrawsocket/afinetrawsocket_windows.go]: fixed in code

File: controller/internal/enforcer/nfqdatapath/datapath_tcp.go (hunk 9 above)
Comment (philipatl) [reply controller/internal/enforcer/nfqdatapath/datapath_tcp.go:390]: you are right. fixed.

File: controller/internal/supervisor/iptablesctrl/acls_windows.go (hunk 13 above)
Comment (philipatl) [reply controller/internal/supervisor/iptablesctrl/acls_windows.go:203]: fixed by right my own Equal comparison

File: controller/internal/supervisor/iptablesctrl/acls_windows.go (hunk 14 above)
Comment (philipatl) [reply controller/internal/supervisor/iptablesctrl/acls_windows.go:186]: I have modified this code to try to make more sense. You still have to check a fair number of conditions, but I hope it makes more sense now. (I'm trying to check that one rule has an Action and the other has Log attributes, and if so then save the appropriate related attributes.)

File: controller/internal/supervisor/iptablesctrl/acls_windows.go (hunk 15 above)
Comment (philipatl) [reply controller/internal/supervisor/iptablesctrl/acls_windows.go:232]: fixed

File: controller/internal/windows/iptables_windows.go (hunk 16 above)
Comment (philipatl) [reply controller/internal/windows/iptables_windows.go:103]: I had assumed that the passed in WinRuleSpec would have been created via a previous parse, which should guarantee that the rulespec is valid. But since this is a public function, I can't guarantee it, so I modified the code by adding a 'validate bool', which if true will do a final parse and check for error. This keeps the 'rulespec validity check' code in one place, too.

File: controller/internal/windows/iptables_windows.go (hunk 17 above)
Comment (philipatl) [reply controller/internal/windows/iptables_windows.go:123]: fixed as above with validity checking

**Comment by itsai220** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "e5508908e26db2d811d5e81b038aa7a33346e4b5",
    "pipeline": "preproduction"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "92302ff6785e51a205e4209060001ed4096d1128",
    "pipeline": "preproduction"
  }
]
```

**Comment by itsai220** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "3b4cffa4416ba0a7fb8d66873798311061b28ace",
    "pipeline": "preproduction"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "92302ff6785e51a205e4209060001ed4096d1128",
    "pipeline": "preproduction"
  }
]
```

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "3b4cffa4416ba0a7fb8d66873798311061b28ace"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "ce9d06e43326bce3920c02300f356ba530bbe929"
  }
]
```

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "3b4cffa4416ba0a7fb8d66873798311061b28ace"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "ce9d06e43326bce3920c02300f356ba530bbe929"
  }
]
```

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "ce9d06e43326bce3920c02300f356ba530bbe929"
  },
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "3b4cffa4416ba0a7fb8d66873798311061b28ace"
  }
]
```

**Comment by itsai220** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "3b4cffa4416ba0a7fb8d66873798311061b28ace",
    "pipeline": "preproduction"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "ce9d06e43326bce3920c02300f356ba530bbe929",
    "pipeline": "preproduction"
  }
]
```

**Review by amitlimaye — DISMISSED** [review]

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "646055deb13ae4bcfda0910d1555a3f5b086ec20"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "dadffd88e4db80f5f387c8a8748f224a6889ac2d"
  }
]
```

File: controller/internal/enforcer/nfqdatapath/nfq_windows.go (hunk 10 above)
Comment (philipatl) [reply controller/internal/enforcer/nfqdatapath/nfq_windows.go:87]: will simplify this code (linux and windows) in another branch and/or after this merge

File: controller/internal/enforcer/nfqdatapath/nfq_windows.go (hunk 11 above)
Comment (philipatl) [reply controller/internal/enforcer/nfqdatapath/nfq_windows.go:111]: will simplify code (linux and windows) in another branch and/or after this merge

File: controller/internal/enforcer/nfqdatapath/nfq_windows.go (hunk 12 above)
Comment (philipatl) [reply controller/internal/enforcer/nfqdatapath/nfq_windows.go:151]: will simplify code (linux and windows) in another branch and/or after this merge

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "6d0462fccc6fd3bd1416ba9d02d785da3ab65193"
  },
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "66c5ab5a0d75443c8d085bba6242222649b7ff8c"
  }
]
```

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "windows2",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "66c5ab5a0d75443c8d085bba6242222649b7ff8c"
  },
  {
    "project": "windows2",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "6d0462fccc6fd3bd1416ba9d02d785da3ab65193"
  }
]
```

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "windows2",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "d968e1c04c91046ed0ed02f0c644e3ba5da1ff44"
  },
  {
    "project": "windows2",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "6d0462fccc6fd3bd1416ba9d02d785da3ab65193"
  }
]
```

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "windows2",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "21032931c096faa62d920af6dd5d5b92316a07ec"
  },
  {
    "project": "windows2",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "ff89f605740352b5679ab0280750deab7408770c"
  }
]
```

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "windows2",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "0ae3be89c2a936133824be6710119aa326eac40d"
  },
  {
    "project": "windows2",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "21032931c096faa62d920af6dd5d5b92316a07ec"
  }
]
```

**Review by dstiliadis — APPROVED** [review]

**Comment by philipatl** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "windows2",
    "component": "trireme-lib",
    "pr-id": "917",
    "commit-sha": "0ae3be89c2a936133824be6710119aa326eac40d"
  },
  {
    "project": "windows2",
    "component": "enforcerd",
    "pr-id": "1498",
    "commit-sha": "f75b033b020d4434148078184e7f18b36d395a02"
  }
]
```
